core/src/main/java/com/segment/analytics/kotlin/core/utilities/OkHttpURLConnection.kt

```diff
@@ -0,0 +1,393 @@
+        }
+        
+        if (requestBodyBuffer == null) {
+            requestBodyBuffer = Buffer()
+        }
+        return requestBodyBuffer!!.outputStream()
+    }
+
+    override fun getHeaderField(name: String?): String? {
+        if (!connected) {
+            try { connect() } catch (e: IOException) { return null }
+        }
+        return response?.header(name ?: return null)
+    }
+
+    override fun getHeaderFields(): MutableMap<String, MutableList<String>> {
+        if (!connected) {
+            try { connect() } catch (e: IOException) { return mutableMapOf() }
+        }
+        return response?.headers?.toMultimap()?.mapValues { it.value.toMutableList() }?.toMutableMap()
+            ?: mutableMapOf()
+    }
+
+    override fun getResponseCode(): Int {
+        if (!connected) connect()
+        return responseCode
+    }
+
+    override fun getResponseMessage(): String {
+        if (!connected) connect()
+        return responseMessage ?: ""
+    }
+
+    /**
+     * Returns the protocol used for this connection (e.g., "h2" for HTTP/2, "http/1.1" for HTTP/1.1).
+     * Returns null if the connection hasn't been established yet.
+     */
+    fun getProtocol(): String? {
+        return response?.protocol?.toString()
+    }
+
+    // Override ALL URLConnection methods to prevent bypassing OkHttp
+    override fun getURL(): URL = url
+
+    override fun getHeaderFieldKey(n: Int): String? {
+        if (!connected) {
+            try { connect() } catch (e: IOException) { return null }
+        }
+        val headers = response?.headers ?: return null
+        return if (n in 0 until headers.size) headers.name(n) else null
+    }
+
+    override fun getHeaderField(n: Int): String? {
+        if (!connected) {
+            try { connect() } catch (e: IOException) { return null }
+        }
+        val headers = response?.headers ?: return null
+        return if (n in 0 until headers.size) headers.value(n) else null
+    }
+
+
+    override fun getPermission(): java.security.Permission? {
+        return java.net.SocketPermission("${url.host}:${if (url.port == -1) url.defaultPort else url.port}", "connect,resolve")
+    }
+
+    override fun setAuthenticator(auth: java.net.Authenticator?) {
+        // OkHttp handles authentication differently, this is a no-op
+    }
+
+    override fun setChunkedStreamingMode(chunklen: Int) {
+        // OkHttp handles chunked encoding automatically, store for compatibility
+    }
+
+    override fun setFixedLengthStreamingMode(contentLength: Int) {
+        // OkHttp handles content length automatically, store for compatibility
+    }
+
+    override fun setFixedLengthStreamingMode(contentLength: Long) {
+        // OkHttp handles content length automatically, store for compatibility
+    }
+
+    override fun setRequestMethod(method: String) {
+        if (connected) {
+            throw IllegalStateException("Already connected")
+        }
+        _requestMethod = method
+    }
```

Comment:
SO internally we can do this (goes from GET to POST), but if someone is using the API tries it should fail; cool.